idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/proofs/markers/valueArgumentCoercion.kt

```diff
@@ -0,0 +1,54 @@
+package arrow.meta.ide.plugins.proofs.markers
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.ide.plugins.proofs.inspections.PairTypes
+import arrow.meta.ide.plugins.proofs.inspections.PairTypes.Companion.pairOrNull
+import arrow.meta.ide.plugins.proofs.inspections.resolveKotlinType
+import arrow.meta.ide.resources.ArrowIcons
+import arrow.meta.phases.CompilerContext
+import arrow.meta.phases.ExtensionPhase
+import arrow.meta.plugins.proofs.phases.areTypesCoerced
+import arrow.meta.plugins.proofs.phases.coerceProof
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtValueArgument
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+val IdeMetaPlugin.implicitCoercionValueArgumentLineMarker: ExtensionPhase
+  get() = addLineMarkerProviderM(
+    icon = ArrowIcons.ICON4,
+    composite = KtValueArgument::class.java,
+    transform = { psiElement: PsiElement ->
+      psiElement.ctx()?.let { ctx ->
```

Comment:
This structure to check for coercion seems to be replicated in different line markers in this PR, maybe we should provide some sugar for it and reuse to avoid duplication?

Reply:
hmmm not sure about that.. hints are welcome! 😄

Reply:
Both `implicitCoercionValueArgumentLineMarker` and `implicitCoercionPropertyLineMarker` files seem to be really similar. Can't we:
* Move both to use `addLineMarkerProviderM` (might need to update the method to support all scenarios but still we are gonna need a common linemarker provider for all our proofs as discussed, so it takes care of merging or not depending on the case, and we can also append clickable options to it etc)
* move all the `.participatingTypes()` extension functions to a common extensions file, `ParticipatingTypesSyntax` or similar (so they can be reused from both files and from everywhere)
* Pull out `isCoerced` to somewhere else and reuse since it is also replicated. Overall we need to find a way to synthesize our apis and leverage reusability.
* Reencode both markers to reuse mostly everything.

Reply:
this has changed today after the refactor.. lets discuss it tomorrow 👍